File: leetcode/java/task1.java
package leetcode.java;

import java.util.Arrays;
import java.util.HashMap;

public class task1 {

    public static int[] twoSum(int[] numbers, int target) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < numbers.length; i++) {
            if (map.containsKey(target - numbers[i]) && i != map.get(target - numbers[i])) {
                return new int[]{i, map.get(target - numbers[i])};
            }
            map.put(numbers[i], i);
        }
        return null;
    }

    public static void main(String[] args) {
        int[] arr = new int[]{3, 2, 4};
        System.out.println(Arrays.toString(twoSum(arr, 6)));
    }
}

File: leetcode/scala/task1.scala
package leetcode.scala

import scala.collection.immutable.HashMap


// неоптимальный
object task1 extends App {
  def twoSum(numbers: Array[Int], target: Int): Array[Int] =
    numbers
      .zipWithIndex
      .foldLeft((HashMap.empty[Int, Int], List.empty[Int])) {
        case ((acc, arr), (num, i)) if acc.contains(target - num) && (!acc.get(target - num).contains(i)) => (acc.updated(num, i), arr ++ List(acc.getOrElse(target - num, 0), i))
        case ((acc, arr), (num, i)) => (acc.updated(num, i), arr)
      }
      ._2
      .toArray
}

// более оптимальный вариант
object task1_opt extends App {
  def twoSum(numbers: Array[Int], target: Int): Array[Int] = {
    val indices = scala.collection.mutable.Map[Int, Int]()

    for (i <- numbers.indices) {
      indices.get(target - numbers(i)) match {
        case Some(j) => return Array(j, i)
        case None => indices(numbers(i)) = i
      }
    }
    Array.empty[Int]
  }
}